### DISCUSSION TITLE
How to properly use TypeScript with react-navigation?

### DISCUSSION THREAD

**Discussion by bviebahn** [react-navigation/react-navigation#10594] [Q&A]
I have been using react-navigation for many years now and I like almost everything about it. The one thing I was never really satisfied with, because I never got it completely right, is using TypeScript types to correctly annotate the navigation and route objects.

Consider this simplified navigation structure, which is probably pretty common: A Root StackNavigator which contains a TabNavigator and some other screens. The TabNavigator contains two StackNavigators, which have some screens that might be unique to that Navigator, but also other screens that might also appear in other Navigators.

```typescript
type HomeStackParams = {
    HomeScreen: { homeScreenParam: string };
    SharedScreen: { sharedScreenParam: string };
};

type ProfileStackParams = {
    ProfileScreen: { profileScreenParam: string };
    SharedScreen: { sharedScreenParam: string };
};

type BottomTabParams = {
    HomeTab: NavigatorScreenParams<HomeStackParams>;
    ProfileTab: NavigatorScreenParams<ProfileStackParams>;
};

type RootStackParams = {
    Login: { loginScreenParam: string };
    Tabs: NavigatorScreenParams<BottomTabParams>;
};
```

One way to use TypeScript with this navigation structure I have come across is to create a type for each navigator:

```typescript
type HomeStackScreenProps<
    Screen extends keyof HomeStackParams
> = CompositeScreenProps<
    BottomTabScreenProps<BottomTabParams, "HomeTab">,
    NativeStackScreenProps<HomeStackParams, Screen>
>;

type ProfileStackScreenProps<
    Screen extends keyof ProfileStackParams
> = CompositeScreenProps<
    BottomTabScreenProps<BottomTabParams, "ProfileTab">,
    NativeStackScreenProps<ProfileStackParams, Screen>
>;

type RootStackScreenProps<
    Screen extends keyof RootStackParams
> = NativeStackScreenProps<RootStackParams, Screen>;
```

Then we could choose between the types depending on the screen we're in, which is not ideal since we have to make changes in multiple places when adding screens that are used in multiple navigators to another navigator, and TypeScript will not help us to enforce changing it in all places. If we make changes to our navigation structure, we always have to think about which screens might be affected, which ideally would be done by TypeScript.

```typescript
const HomeScreen = () => {
    // ok if we assume that this screen will never be used outside the Home StackNavigator
    const navigation = useNavigation<
        HomeStackScreenProps<"SharedScreen">["navigation"]
    >();
    return null;
};

const SharedScreen = () => {
    // not ideal since we have to keep track where this screen will be used.
    // If we decide to add it to the RootStack there will be no error here.
    const navigation = useNavigation<
        | HomeStackScreenProps<"SharedScreen">["navigation"]
        | ProfileStackScreenProps<"SharedScreen">["navigation"]
    >();
    return null;
};
```

Ideally, I would like to define my navigation structure once, so that I don't have to make any assumption inside specific screens about where it will be used, and have it enforced by TypeScript everywhere. I tried to create a single type that correctly models the whole navigation structure. I'm not sure if this is actually possible. Every few months or so I come back to this problem and play around a little bit, but I was never able to solve it, but maybe someone with better TypeScript skills has a solution. I was thinking of something like this (but different, since this doesn't work at all):

```typescript
type NavigationProps<
    Screen extends
        | keyof RootStackParams
        | keyof BottomTabParams
        | keyof HomeStackParams
        | keyof ProfileStackParams
> = CompositeScreenProps<
    RootStackScreenProps<
        Screen extends keyof RootStackParams ? Screen : keyof RootStackParams
    >,
    CompositeScreenProps<
        BottomTabScreenProps<
            TabsParams,
            Screen extends keyof TabsParams ? Screen : keyof TabsParams
        >,
        | (Screen extends keyof HomeStackParams
              ? HomeStackScreenProps<Screen>
              : never)
        | (Screen extends keyof ProfileStackParams
              ? ProfileStackScreenProps<Screen>
              : never)
    >
>;
```

Any tips on how to achieve better type safety are appreciated!

**Comment by satya164** [comment]
https://reactnavigation.org/docs/typescript/#organizing-types

**Reply by bviebahn** [reply]
The method that's used in the docs is a bit flawed imo as I described above.
```typescript
function PopularScreen({ navigation, route }: HomeTabScreenProps<'Popular'>) {
  // ...
}
```
If I decide to include the "Popular" screen in another Navigator I have to update the type in two places. For example I might want to add the PopularScreen to a different Tab, let's say "Profile". I add the PopularScreen to the "ProfileStackParams", and render it in the Navigator. At this point, TypeScript is happy and there will be no errors. However, PopularScreen might use the navigation object to navigate to a screen that's only available in the HomeTab stack. So I have to add that screen to the ProfileTab stack as well, and ideally, TypeScript would force me to do that.

**Reply by satya164** [reply]
if you're using the same screen in multiple navigators then it's tricky to type the navigation prop - it should be `FirstNavigatorScreenProps<'Popular'> | SecondNavigatorScreenProps<'Popular'>` - but TS won't be happy when using it.

however, you can use `useNavigation` which will work anywhere given that you provided a global type.

**Reply by bviebahn** [reply]
I think this `FirstNavigatorScreenProps<'Popular'> | SecondNavigatorScreenProps<'Popular'>` works fine, but I still have to keep track of where the screen is used, so it kind of breaks modularity if you know what I mean. And I think it is very common for screens to be used in multiple navigators, right?

> however, you can use useNavigation which will work anywhere given that you provided a global type.

Not sure I understand, doesn't this have the same problems?

Maybe this is the best way at the moment (and since I don't see many others complaining about it it's probably good enough for most people), but I think it could be better and it should be possible to infer the type in the specific screens without having to make assumptions on where exactly it will be used.

Since this is such a popular library maybe there is some TypeScript genius able to improve this.

**Reply by satya164** [reply]
> Not sure I understand, doesn't this have the same problems?

How does it have the same problem? You define the params for `useNavigation` once and it works the same everywhere

**Reply by satya164** [reply]
> but I think it could be better and it should be possible to infer the type in the specific screens without having to make assumptions on where exactly it will be used.

It's not possible infer how things are nested, but feel free to send a PR or open an RFC if you figure out any improvements

**Reply by bviebahn** [reply]
Thanks, I didn't know about the RFC process. I doubt I could implement the types myself, but I'll think about it some more and then probably write a RFC when I have time.

**Reply by MathisBarre** [reply]
Hello @bviebahn ! I came to the same conclusion as you. So, if i understand well your discussions with @satya164, the simplest method is to only use useNavigation with only one type for all the application navigators ?

**Reply by bviebahn** [reply]
@MathisBarre Unfortunately I haven't found a perfect solution yet. I'm just trying to make sure I update the correct types everywhere when I change something related to the navigation, and I keep in mind that unfortunately I can't fully rely on Typescript to cover for every mistake I could make.
